Implement discrete replicator dynamics (#236)
drvinceknight/Nashpy: docs/text-book/discrete-replicator-dynamics.rst

.. _discrete-replicator-dynamics:

Discrete Replicator dynamics
============================
This is the discrete version of the continuous :ref:`replicator dynamics <replicator-dynamics-discussion>` and can be used to model how stratergies evolve over time
in distinct increments.

.. _motivating-example-discrete_replicator_dynamics:

consider the Hawk Dove model shown in :ref:`replicator dynamics <replicator-dynamics-discussion>`, where we wish to model the populations of an aggressive animal and sharing animal over time

> but assume the population of aggressive and sharing animals only changes once a day when they interact to eat. 

Reviewer: ```suggestion
Discrete Replicator dynamics
============================

Motivating example: once a day hawk dove game
---------------------------------------------

Consider the Hawk Dove model shown in :ref:`replicator dynamics <replicator-dynamics-discussion>`, where we wish to model the populations of an aggressive animal and sharing animal over time
but assume the population of aggressive and sharing animals only changes once a day when they interact to eat. 
```

we can use the same payoff matrix

> :math:`A`:

Reviewer: ```suggestion
We use the payoff matrix :math:`A`:
```

.. math::

   A = \begin{pmatrix}
       2 & 1\\
       3 & 0
   \end{pmatrix}

and starting population distribution

:math:`x_0 = ([0.05,0.95])`:
where 95% are sharers while 5% are aggressive

> we can use the following discrete function to find the new population after each time step

Reviewer: ```suggestion
We can use the following discrete function to find the new population after each time step:
```

:math:`x_{\ i+1} = x_i (\frac{(Ax)_i}{x^T A x})`:

> and get discrete results (on the left) similar to the continuous replicator function (on the right)

Reviewer: ```suggestion
and get discrete results (on the left) similar to the continuous replicator function (on the right).
```

   discrete_sharing_population, discrete_aggressive_population = game.discrete_replicator_dynamics(y0, steps=20).T

   fig, (ax1, ax2) = plt.subplots(1, 2)

   ax1.set_xticks(np.arange(0, 21, step=2))
   ax1.set_ylim(0, 1)
   ax2.set_xticks(np.arange(0, 21, step=2))
   ax2.set_ylim(0, 1)
   ax1.step(np.arange(0, 21, step=1),np.append(y0[0],discrete_sharing_population),label="sharer")
   ax1.step(np.arange(0, 21, step=1),np.append(y0[1],discrete_aggressive_population),label="aggressor")
   ax2.plot(cont_linspace,sharing_population,label="sharer")
   ax2.plot(cont_linspace,aggressive_population,label="aggressor")
   ax1.legend() 
   ax2.legend() 

both the continuous and discrete replicator functions produce a proportion of a total population,
if there were a total of 100 animals in this example this would result in non-intager animal populations.

to resolve this we can use greenwoods algorithm after each discrete time step. 

based on the algorithm described by [Greenwood2019]_,
the quantiszation algorithm rounds the population of each stratergy to its nearest intager value after each step, 

> then increments or decrements values with the largest error to ensure the total population stays consistent.

Reviewer: ```suggestion
Both the continuous and discrete replicator functions produce a proportion of a total population, this addresses
the first modelling challenge: have discrete time.

However what if we wanted to model a finite population of 100 animals?

Greenwoods Quantization Algorithm
---------------------------------

This algorithm is described in [Greenwood2019]_.

TODO Describe the algorithm rigorously.

```

   quantize_sharing_population, quantize_aggressive_population = game.discrete_replicator_dynamics(y0q, steps=20,quantize=True).T

   fig, (ax1, ax2) = plt.subplots(1, 2)

   ax1.step(np.arange(0, 21, step=1),np.append(y0[0],discrete_sharing_population),label="sharer")
   ax1.step(np.arange(0, 21, step=1),np.append(y0[1],discrete_aggressive_population),label="aggressor")
   ax2.step(np.arange(0, 21, step=1),np.append(y0q[0],quantize_sharing_population),label="sharer")
   ax2.step(np.arange(0, 21, step=1),np.append(y0q[1],quantize_aggressive_population),label="aggressor")

   ax1.set_xticks(np.arange(0, 21, step=2))
   ax1.set_ylim(0, 1)
   ax2.set_xticks(np.arange(0, 21, step=2))
   ax2.set_ylim(0, 100)

   ax1.legend() 
   ax2.legend() 

Discussion
----------

due to the requirement that the population must stay consistent, the quantisation algorithm can reach artificial fixed points.
These occur when the discrete replicator algorithm would produce a change in population distribution
though this change is small enough to be undone when rounded. This can be seen above where the quantized algorithm does not reach a 50/50
distribution of aggressive and sharing animals.

> 

Reviewer: ```suggestion

```

Let's remove this as I think it needs more careful consideration.

.. _replicator-dynamics-discussion:

Discrete Replicator dynamics

> ===================

Reviewer: ```suggestion
Discrete Replicator dynamics
============================

```

This is considered as the discrete version of the continuous :ref:`replicator-dynamics` and can be used to model how stratergies evolve over time

> in systems that dont neccicarily change contiuously over time but rather in distinct increments.

Reviewer: ```suggestion
in systems that doesn't necessarily change continuously over time but rather in distinct increments.
```
